### PR TITLE
.*: fix bug that after execute pause-task the task may still running 

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
### What problem does this PR solve? 

fix issue https://github.com/pingcap/dm/issues/637 and https://github.com/pingcap/dm/issues/629

when executing pause task the task is not real paused, and then execute resume task, maybe the old unit is still process, and then will cause some problem.

### What is changed and how it works?

1. add two stage `pausing` and `resuming`:
- when pausing task, first the stage is `pausing`, and update stage to `paused` at the last.
- when resuming task, first the stage is `resuming`, and update stage to `running` at the last.

2. a minor fix, `fetchResult` should wait for the unit's process result.

### Check List 

Tests 

 - Integration test

### CODE REVIEW

File: dm/worker/subtask.go [hunk 1]
```diff
@@ -236,18 +236,16 @@ func (st *SubTask) callCurrCancel() {
 func (st *SubTask) fetchResult(pr chan pb.ProcessResult) {
 	defer st.wg.Done()
 
-	st.RLock()
-	ctx := st.currCtx
-	st.RUnlock()
-
 	select {
-	case <-ctx.Done():
+	case <-st.ctx.Done():
```
Comment (csuzhangxc) [inline dm/worker/subtask.go:240]: Should we need to fix this in release-1.0? cc @lichunzhu

File: dm/unit/unit.go [hunk 2]
```diff
@@ -77,3 +78,11 @@ func NewProcessError(err error) *pb.ProcessError {
 	}
 	return result
 }
+
+// IsCtxCanceledProcessErr returns true if the err's
```
Comment (csuzhangxc) [inline dm/unit/unit.go:82]: Do some comments lost at the end?

File: dm/worker/subtask.go (hunk 1 above)
Comment (lichunzhu) [reply dm/worker/subtask.go:240]: I think it's not correct. We will start multi `fetchResult` in the end. `fetchResult` should be ended when one turn is finished.

Comment (csuzhangxc) [reply dm/worker/subtask.go:240]: but we still waiting for `case result := <-pr:`, is that enough? @lichunzhu

File: dm/worker/subtask.go [hunk 3]
```diff
@@ -462,6 +472,7 @@ func (st *SubTask) Resume() error {
 		st.setStage(pb.Stage_Paused)
 		return err
 	} else if ctx.Err() != nil {
+		st.setStage(pb.Stage_Paused)
```
Comment (lichunzhu) [inline dm/worker/subtask.go:475]: I think this should not be set to `pb.Stage_Paused`, see https://github.com/pingcap/dm/pull/589#discussion_r409342791

File: dm/worker/subtask.go (hunk 1 above)
Comment (lichunzhu) [reply dm/worker/subtask.go:240]: It's enough, and we mush assure `pr` will get a result in the future.

File: dm/worker/subtask.go (hunk 3 above)
Comment (WangXiangUSTC) [reply dm/worker/subtask.go]: ok, I remove this and add a comment e34ec49

File: tests/sharding/run.sh [hunk 4]
```diff
@@ -90,12 +90,21 @@ function run() {
...
     # wait really paused
-    # FIXME: `if !st.stageCAS(pb.Stage_Running, pb.Stage_Paused)` in `subtask.go` is not enough to indicate the real stage.
-    sleep 2
+    run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+        "query-status test" \
+        "Paused" 2
+    run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+        "resume-task test"\
+        "\"result\": true" 3
+    # resume twice
```
Comment (lichunzhu) [inline tests/sharding/run.sh:107]: Should we add a test to test `stop-task` twice?

Comment (WangXiangUSTC) [reply tests/sharding/run.sh]: add in ecd45e6

**Pull request merged**